### ISSUE TITLE
ESLint 6.2.0 + babel-eslint + no-unused-vars false positive with for-in loop

### ISSUE THREAD

**Issue by feross** [eslint/eslint#12117]
There's a new issue in ESLint 6.2.0 just caught by the `standard` test suite. Issue did not exist in ESLint 6.1.0.

**Tell us about your environment**

* **ESLint Version:** 6.2.0
* **Node Version:**10.16.3
* **npm Version:** 6.9.0

**What parser (default, Babel-ESLint, etc.) are you using?** babel-eslint

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
Config:

```
{
  "parserOptions": {
    "ecmaVersion": 2019,
    "ecmaFeatures": {
      "jsx": true
    },
    "sourceType": "module"
  },
  "parser": "babel-eslint",
  "rules": {
    "no-unused-vars": ["error", { "vars": "all", "args": "none", "ignoreRestSiblings": true }]
  }
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

<!-- Paste the source code below: -->
```js
function scoreAudioCoverFile (imgFile) {
  const fileName = path.basename(imgFile.name, path.extname(imgFile.name)).toLowerCase()
  const relevanceScore = {
    cover: 80,
    folder: 80,
    album: 80,
    front: 80,
    back: 20,
    spectrogram: -80
  }

  for (const keyword in relevanceScore) {
    if (fileName === keyword) {
      return relevanceScore[keyword]
    }
    if (fileName.indexOf(keyword) !== -1) {
      return relevanceScore[keyword]
    }
  }
  return 0
}
```

```bash
npx eslint --config eslintrc.json t.js
```

**What did you expect to happen?**

Just one error:

```
  1:10  error  'scoreAudioCoverFile' is defined but never used  no-unused-vars
```

**What actually happened? Please include the actual, raw output from ESLint.**

Two errors:

```
   1:10  error  'scoreAudioCoverFile' is defined but never used  no-unused-vars
  12:14  error  'keyword' is defined but never used              no-unused-vars
```

The variable `keyword` from the for-in loop is seen as not used.

This may be a bug in `babel-eslint` since the issue does not happen when the default parser is used. However, this issue did not exist with `babel-eslint` + 6.1.0 and does exist with `babel-eslint` + 6.2.0.

**Are you willing to submit a pull request to fix this bug?**
Yes

**Comment by h4emp3** [comment]
I just noticed this as well, maybe a second example might help in finding the problem.

test.js
```js
export const foo = bar => {
    const baz = {};
    for (const key of bar) {
        baz[key] = [];
    }
    return baz;
};
```

```sh
eslint --parser babel-eslint --rule '{"no-unused-vars": ["error"]}' --no-eslintrc test.js
```

**Comment by feross** [comment]
This may have been caused by the change to the `no-unused-vars` rule introduced in this PR: https://github.com/eslint/eslint/pull/12055 cc @mdjermanovic

**Comment by h4emp3** [comment]
Well, at least the rule doesn't result in a false positive for my example when I just add the additional condition again in a otherwise unchanged v6.2.0.

**Comment by tmorehouse** [comment]
We are getting man false positives with 6.2.0 as well:

> 6:14  error  'arg' is defined but never used   no-unused-vars

```js
  for (const arg of args) {
    if (arg !== undefined) {
      return arg
    }
  }
```

>  111:16  error  'type' is defined but never used  no-unused-vars

```js
    for (const type of ['a', 'b']) {
      const x = Array.isArray(options[type]) ? options[type] : []
```

**Comment by mdjermanovic** [comment]
`babel-eslint` is using `"eslint-scope": "3.7.1"`

**Comment by mdjermanovic** [comment]
> This may have been caused by the change to the `no-unused-vars` rule introduced in this PR: #12055 cc @mdjermanovic

Most likely it is that commit, with `babel-eslint` in the keyword example there is a `TDZ` scope between `function` and `for`, probably because it's using the old version of `eslint-scope`

**Comment by kaicataldo** [comment]
@feross Any chance you can test the same code out with the default parser to determine if this issue should be fixed in babel-eslint or here? @mdjermanovic's assessment is most likely what's going on here.

**Comment by hellomrbigshot** [comment]
The same question! I use 6.1.0 instead and it works.

**Comment by feross** [comment]
@kaicataldo The issue doesn't exist with the default parser.

**Comment by aladdin-add** [comment]
thanks for the feedback!

technically this is not an issue in eslint; but to avoid blocking users, **personally**(not for the team) I'd suggest to revert that change and release a patch. 

we can reintroduce it when it was fixed in babel-eslint. :)

**Comment by andsav** [comment]
This issue was introduced in 6.2.0. For example this code:

```
  for (const folder of splitPath) {
      if ('' in current) {
        return
      }
      current = current[folder]
    }
```

throws an eslint error:

`no-unused-vars:  'folder' is defined but never used`

**Comment by tacoto12** [comment]
She needs to call me

**Comment by jharris4** [comment]
I just hit the same issue with a bunch of false positives after upgrading to `6.2.0`. I'm also using `babel-eslint`.

Each of the false positives for the `no-unused-vars` that I'm seeing are related to variable declared in a for...of loop, like:

```
for (let theVariable of theCollection)
```

This triggers an error for `theVariable`.

**Comment by Bobsans** [comment]
+1

**Comment by kaicataldo** [comment]
> technically this is not an issue in eslint; but to avoid blocking users, personally(not for the team) I'd suggest to revert that change and release a patch.

I'm not sure reverting the change is the best option.

Due to our [semver policy](https://github.com/eslint/eslint/#semantic-versioning-policy), we advise that `~` is used instead of `^`, since minor versions can include bug fixes that can increase the number of errors.

Locking the ESLint version or using `~6.1.0` should make sure people can continue using babel-eslint until babel-eslint fixes the issue.

**Comment by kaicataldo** [comment]
@feross Thanks for confirming! I see there's already [an issue open for babel-eslint](https://github.com/babel/babel-eslint/issues/791), and it looks like we have a lead on what might be causing it. I think we should leave this issue open until it's fixed in babel-eslint so that people who encounter this error find the temporary workaround, but can discussion on how to fix it move to that issue?

**Comment by vincerubinetti** [comment]
@kaicataldo I'd leave it open for discussion. As of writing, it's the first result on Google for this error.

**Comment by kaicataldo** [comment]
@vincerubinetti I think you and I just said the same thing 😄

**Comment by przemyslawzalewski** [comment]
babel-eslint is using an outdated version of [eslint-scope](https://github.com/babel/babel-eslint/blob/c7e3817f3ee64801eb30c53e575099dc089cd915/package.json#L36) and doesn't really support ESLint 6 as the ESlint requires [eslint-scope@^5.0.0](https://github.com/eslint/eslint/blob/4aeeeedb656ee3519ea82ebf0cb41ca801215046/package.json#L52) but babel-eslint requires [a pinned version of 3.7.1](https://github.com/babel/babel-eslint/blob/c7e3817f3ee64801eb30c53e575099dc089cd915/package.json#L36).
I have managed to [fix the issue by updating the dependencies](https://github.com/babel/babel-eslint/pull/792) and [including proposed changes to match Espree 6 AST](https://github.com/babel/babel-eslint/pulls). This fork is based on a not yet released major version of babel-eslint and may come with breaking changes.
Anyway, the fork (which can be installed by just replacing the regular version inside package.json with `"babel-eslint": "github:przemyslawzalewski/babel-eslint#b6965ab8dedd5a9374475e692912a0df2f239262"`) works for me and yields no false positives so I will wait for the babel-eslint to be released with the fixes.

**Comment by przemyslawzalewski** [comment]
I have proposed [another PR as a PATCH change](https://github.com/babel/babel-eslint/pull/793) for the babel-eslint repository that should be easier to integrate and I hope it will be merged soon.
In the meanwhile, I have released a forked package for the ^10 version with the fixes included:

https://www.npmjs.com/package/@przemyslawzalewski/babel-eslint/v/10.0.3

Usage is simple as changing some entires in a package.json:
`"babel-eslint": "^10.0.2"` -> `"@przemyslawzalewski/babel-eslint": "^10.0.3"` and 
`"parser": "babel-eslint"` -> `"parser": "@przemyslawzalewski/babel-eslint"` within the eslint config file.

There is also a fork of the not yet released ^11 branch, however, it may include some breaking changes:
https://www.npmjs.com/package/@przemyslawzalewski/babel-eslint/v/11.0.0

**Comment by Sun79** [comment]
You can also solve this problem by adding the following code to package.json.
```json
"resolutions": {
  "babel-eslint/eslint-scope": "^5.0.0"
}
```

**Comment by ljharb** [comment]
@Sun79 that's not something npm supports.

**Comment by yoyo837** [comment]
@Sun79 Only yarn works.

**Comment by Sun79** [comment]
@ljharb @yoyo837 Sorry, I made a mistake.😭

**Comment by benwoodward** [comment]
I just experienced this issue, here's another example: 

```javascript
  for (const [fileType, fileExt] of Object.entries(extensions)) {
    try {
      componentFiles[fileType] = dir + "/" + name + fileExt;
    } catch (e) {}
  }
```

**Comment by yoyo837** [comment]
https://github.com/babel/babel-eslint/pull/794 fixed this.

**Comment by yoyo837** [comment]
@devinrhode2 upgrade `babel-eslint` to `v10.0.3`.

**Comment by shubhranshu** [comment]
I got eslint 5.15 and it works for me.

**Comment by marcinlesek** [comment]
Any updates here?

**Comment by EmilyRosina** [comment]
For me it has been fixed :woman_shrugging: we're on the latest version with no problems

**Comment by tacoto12** [comment]
I am going to come get rosa mae PERIOD!!

**Comment by vincerubinetti** [comment]
@tacoto12 wat

**Comment by marcinlesek** [comment]
Hmm, I'm on `6.3.0` and it's still occurring for me, unfortunately. Should I create another ticket?

**Comment by platinumazure** [comment]
@marcinlesek Can you try blowing away your node_modules and reinstalling? And if the issue recurs, please then run `npm ls eslint-scope` and see if you can find if there's an old eslint-scope somewhere-- if so, see if the dependency which contains the old version of eslint-scope can be upgraded.

At this point, the root cause is likely npm's package deduplication logic, not anything on the ESLint side. (But please let us know if you think differently after investigating.)

**Comment by jharris4** [comment]
@marcinlesek It's also working for me since upgrading `babel-eslint` to latest. What version of `babel-eslint` are you running?

**Comment by tacoto12** [comment]
Bold mf

**Comment by marcinlesek** [comment]
@platinumazure @jharris4 both answers are correct guys, thanks! I upgraded `eslint-scope` from `3.7.1` to `5.0.0` and babel aslant from `10.0.1` to `10.0.3` and it now works correctly. 

Kudos!

**Comment by dbroadhurst** [comment]
"eslint": "6.3.0",
    "eslint-config-standard": "14.1.0",
    "eslint-plugin-import": "2.18.2",
    "eslint-plugin-node": "10.0.0",
    "eslint-plugin-promise": "4.2.1",
    "eslint-plugin-standard": "4.0.1",

Gives me an unused warning on test

````
const tests=['a']

for (let test of tests) {
  console.log(test)
}
````

**Comment by platinumazure** [comment]
@dbroadhurst Please share your ESLint configuration and your actual ESLint version. Help us to help you :smile:

**Comment by dbroadhurst** [comment]
I've added my .eslintrc file. Not sure what you mean by actual ESLint version.

````
node_modules/.bin/eslint -v
v6.3.0
````

**Comment by dbroadhurst** [comment]
Slight mix up but I have an easy test case. Might be create-react-app issue

````
create-react-app test
````

add the test code and get the warning. eslint is v6.3.0

**Comment by vincerubinetti** [comment]
Idk if it's related, but note that `create-react-app` still hasn't been updated with the fixed version of `eslint`/`babel-eslint`/`eslint-scope`/whatever. The patch has been merged but it hasn't been released and pushed to npm yet, it looks like:

https://github.com/facebook/create-react-app/pull/7662
https://github.com/facebook/create-react-app/issues/7566

**Comment by colemars** [comment]
```
  for (const item of arr) {
    newArr.push(item);
  }
```
Throwing false positive `no-unused-vars` error. I've updated `eslint` and` babel-eslint` but still hitting it. Is it a dependency issue in my case?

```
"devDependencies": {
    "babel-eslint": "^10.0.3",
    "eslint": "^6.4.0",
    "eslint-config-airbnb": "^18.0.1",
    "eslint-config-prettier": "^6.3.0",
    "eslint-config-react": "^1.1.7",
    "eslint-plugin-import": "^2.18.2",
    "eslint-plugin-jsx-a11y": "^6.2.3",
    "eslint-plugin-prettier": "^3.1.0",
    "eslint-plugin-react": "^7.14.3",
    "eslint-plugin-react-hooks": "^1.7.0",
    "prettier": "^1.18.2"
  }
```
```
module.exports = {
  env: {
    es6: true,
    browser: true
  },
  extends: [
    "airbnb",
    "plugin:react/recommended",
    "prettier",
  ],
  globals: {
    Atomics: 'readonly',
    SharedArrayBuffer: 'readonly',
  },
  parserOptions: {
    ecmaVersion: 2018,
    ecmaFeatures: {
      experimentalObjectRestSpread: true,
      jsx: true
  },
    sourceType: 'module',
  },
  rules: {
    "prettier/prettier": "error",
    "react/react-in-jsx-scope": "off",
    "react/jsx-filename-extension": [1, { "extensions": [".js", ".jsx"] }],
    "jsx-a11y/anchor-is-valid": [ "error", {
      "components": [ "Link" ],
      "specialLink": [ "hrefLeft", "hrefRight" ],
      "aspects": [ "invalidHref", "preferButton" ]
    }]
  },
  plugins: [
    "prettier", "react"
  ],
  settings: {
    react: {
      pragma: "React",
      version: "detect"
    }
  },
  parser: "babel-eslint"
};
```

**Comment by platinumazure** [comment]
> Looks like webpack uses `eslint-scope@4.0.3`. Could that be causing issues?

I think so. Please see if a new version of webpack is available which uses a more recent `eslint-scope`, otherwise you might wish to install eslint-scope directly as a devDependency.

To be clear, this is completely an npm bug, not an ESLint bug. In your dependency structure, ESLint is dependent on latest eslint-scope, but for some reason, npm laid out the physical folder structure in such a way that ESLint gets an old instance of eslint-scope when the linter is run.

**Comment by vincerubinetti** [comment]
To whom it affects:

As of yesterday (I think), `create-react-app` finally [released](https://github.com/facebook/create-react-app/releases) its [fix](https://github.com/facebook/create-react-app/pull/7662) for this `babel-eslint` bug in `create-react-app` version `3.1.2`.

If you were doing any dirty hacks to temporarily get around this, now's the time you can finally remove them.

**Comment by GerardRodes** [comment]
> > Looks like webpack uses `eslint-scope@4.0.3`. Could that be causing issues?
> 
> I think so. Please see if a new version of webpack is available which uses a more recent `eslint-scope`, otherwise you might wish to install eslint-scope directly as a devDependency.
> 
> To be clear, this is completely an npm bug, not an ESLint bug. In your dependency structure, ESLint is dependent on latest eslint-scope, but for some reason, npm laid out the physical folder structure in such a way that ESLint gets an old instance of eslint-scope when the linter is run.

same happens with yarn

solved by upgrading babel-eslint to 10.0.3

**Comment by allan2coder** [comment]
It's a bug in eslint, just upgrade eslint to v6.7.1.

**Comment by MarkPare** [comment]
I upgraded to eslint v6.7.2 and still seeing this.

**Comment by platinumazure** [comment]
> It's a bug in eslint, just upgrade eslint to v6.7.1.

@allan2coder This is not accurate, or at least not fully accurate. babel-eslint did have to write a fix on their side as well. Please ensure you have latest babel-eslint.

> I upgraded to eslint v6.7.2 and still seeing this.

@MarkPare See above: You do need to upgrade babel-eslint to latest as well. If you upgrade babel-eslint and still run into this issue, please open a new issue and fill out our issue template so we can take a look at your configuration and help you more effectively. Thanks!

**Comment by jgreen210** [comment]
...i.e. I also had to remove the .eslintcache directory.

**Comment by lffg** [comment]
I am using ESLint v6.7.2 and this error is happening.

**Comment by RobTheFiveNine** [comment]
If anyone is still running into this issue with React Native due to @react-native-community/eslint-config having a hard reference to version 10.0.1 of babel-eslint and you're using yarn, add this to your `package.json`, remove `node_modules` and re-run `yarn`:

```json
  "resolutions": {
    "@react-native-community/eslint-config/babel-eslint": "^10.0.3"
  }
```
